### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by Chincoya**
Merge implementation of Ruby's enumerable functions.

### CODE REVIEW

File: enumerable.rb [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    if block_given?
+      ar = []
+      my_each do |x|
+        ar << x if yield x
+      end
+    else
+      to_enum
+    end
+    ar
+  end
```
Comment (whiz25) [inline enumerable.rb]: - [ ] returns an enumerator if no block is given. you can achieve this by doing ```return to_enum```

File: enumerable.rb [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+      my_each do |x|
+        return false unless yield x
+      end
+      true
+    else
+      return my_all? { |obj| obj } unless pattern
+
+      my_all? { |x| x.is_a? pattern }
+    end
+  end
```
Comment (whiz25) [inline enumerable.rb]: - [ ] when a class is passed as an argument returns true if all of the collection is a member of such class
- [ ] when a Regex is passed as an argument returns true if all of the collection matches the Regex
- [ ] when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

File: enumerable.rb [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+  def my_any?
+    return my_any? { |obj| obj } unless block_given?
+
+    my_each do |x|
+      return true if yield x
+    end
+    false
+  end
```
Comment (whiz25) [inline enumerable.rb]: - [ ] when a class is passed as an argument returns true if at least one of the collection is a member of such class
- [ ] when a Regex is passed as an argument returns false if none of the collection matches the Regex
- [ ] when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern

File: enumerable.rb [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+  def my_none?
+    return !my_any? unless block_given?
+
+    !my_any? { |x| yield x }
+  end
```
Comment (whiz25) [inline enumerable.rb]: - [ ] when a class is passed as an argument returns true if none of the collection is a member of such class
- [ ] when a Regex is passed as an argument returns true only if none of the collection matches the Regex
- [ ] when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern

File: enumerable.rb [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+        sym = initial
+        initial = self[0]
+      end
+      my_each do |item|
+        next if initial == item
+
+        initial = initial.send sym, item
+      end
+    end
```
Comment (whiz25) [inline enumerable.rb]: - [ ] combines all elements of enum by applying a binary operation, specified by a block

Comment (Chincoya) [reply enumerable.rb]: Hi, Whiz, 
This method already does that. Documentation also included the case of being given a symbol, which is the second part of the method

File: enumerable.rb [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+        my_all? { |x| pattern.match(x) }
+      elsif pattern.class == Class
+        my_all? { |x| x.is_a? pattern }
+      else
+        my_all? { |x| x === pattern }
+      end
+    end
+    my_each { |x| return false unless yield x }
+    true
+  end
```
Comment (whiz25) [inline enumerable.rb]: - [ ] when a class is passed as an argument returns true if all of the collection is a member of such class
- [ ] when a Regex is passed as an argument returns true if all of the collection matches the Regex
- [ ] when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

File: enumerable.rb [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+        my_any? { |x| pattern.match(x) }
+      elsif pattern.class == Class
+        my_any? { |x| x.is_a? pattern }
+      else
+        my_any? { |x| x === pattern }
+      end
+    end
+    my_each { |x| return true if yield x }
+    false
+  end
```
Comment (whiz25) [inline enumerable.rb]: - [ ] when a class is passed as an argument returns true if at least one of the collection is a member of such class
- [ ] when a Regex is passed as an argument returns false if none of the collection matches the Regex
- [ ] when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern

File: enumerable.rb [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+  def my_none?(pattern = nil)
+    if block_given?
+      !my_any? { |x| yield x }
+    elsif pattern
+      !my_any?(pattern)
+    else
+      my_each { |x| return false if x }
+      true
+    end
+  end
```
Comment (whiz25) [inline enumerable.rb]: - [ ] when a class is passed as an argument returns true if none of the collection is a member of such class
- [ ] when a Regex is passed as an argument returns true only if none of the collection matches the Regex
- [ ] when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern

File: enumerable.rb [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+        initial = self[0]
+      end
+      my_each do |item|
+        next if initial == item
+
+        initial = initial.send sym, item
+      end
+    end
+    initial
+  end
```
Comment (whiz25) [inline enumerable.rb]: - [ ] combines all elements of enum by applying a binary operation, specified by a block

**Comment by Chincoya** [comment]
Hello Martin,

Could you please let me know the testing methor you're using? I've tested all those cases and they work. Also, my_inject already does the specified requirement.

Best regards,
Daniel Chincoya

**Comment by whiz25** [comment]
Hey @Chincoya 
I usually test these methods from the terminal(```irb```). There might be no problem with how you test your methods but the problem with some testing methods is that they don't fully cover the edge cases for the methods hence throwing errors when someone tests otherwise.

#### Check how I tested your ```my_all?``` method. I passed a regex it failed.
my_all

#### Now ```my_inject```
my_inject

#### Compare that against ```inject```
inject

#### I am not saying the way you test is wrong, but I am saying your missing some edge cases in your tests. Please check the official [docs](https://ruby-doc.org/core-2.6.5/Enumerable.html) to learn more.

**Comment by Chincoya** [comment]
This should work now, but I have a problem. Stickler is complaining about percieved complexity of methods and module length. I could use auxiliary methods to solve the complexity problem, but that way I couldn't solve the module length one. What should I do?
